src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,361 @@
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Any, Any}, copycols::Bool)
+    newname = last(last(nc))
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    col_idx = first(nc)
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(last(nc)).(df[!, col_idx])
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+        if length(col_idx) == 0
+            newdf[!, newname] = [first(last(nc))() for _ in axes(df, 1)]
+        else
+            cols = ntuple(i -> _columns(df)[col_idx[i]], length(col_idx))
+            colnames = ntuple(i -> _names(df)[col_idx[i]], length(col_idx))
+            newdf[!, newname] = select_transform_helper(TypeHolder{NamedTuple{colnames,
+                                                                              Tuple{eltype.(cols)...}}}(),
+                                                        cols, first(last(nc)), nrow(df))
+        end
+    else
+        throw(ErrorException("code should never reach this branch"))
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing
+provided that the columns requested in each of them are unique and present in `df`.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
+The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`new_column_name` must be a `Symbol`.
+If `old_column` is a `Symbol` or an integer then and `fun` must be callable
+that is applied row by row to the values of `old_column`.
+Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
+will be passed `DataFrameRows` selected by `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `olcd_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then duplicates are accepted except for
+column renaming and transformation operations, where it is not alloweded to rename/transform
+into the same column name.
+For example if `:col` is present in `df` a call `select!(df, :col, :)` is valid
+and moves the column `:col` moved to be the first one in-place.
+
+Note that including the same column several times in the data frame will create aliases.
```

Comment:
Maybe `copycols` should be supported here to avoid this by default? We could copy all renamed columns by default but that would be wasteful, or only in case of duplicates.

Reply:
The only way you can get a duplicate is:
`select(df, :a, :a => :b, :a => :c, copycols=false)` and you have three aliases of the same column produced.
Otherwise we guarantee not to produce duplicates (if you set `copycols=true` you do not have duplicates).
I will clarify the docstring.

Do you think I should add detection of these cases and perform a copy even if `copycols=false`?